services/QuillLMS/app/services/classlink_integration/enrollment_persister.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module ClasslinkIntegration
+  class EnrollmentPersister < ApplicationService
+    attr_reader :incoming_enrollment, :existing_provider_users, :existing_provider_classrooms
+
+    COTEACHER = ClassroomsTeacher::ROLE_TYPES[:coteacher]
+    OWNER = ClassroomsTeacher::ROLE_TYPES[:owner]
+
+    def initialize(incoming_enrollment:, existing_provider_users:, existing_provider_classrooms:)
+      @incoming_enrollment = incoming_enrollment
+      @existing_provider_users = existing_provider_users
+      @existing_provider_classrooms = existing_provider_classrooms
+    end
+
+    def user_and_classroom
+      quill_user = existing_provider_users.find do |ec|
+        ec.external_id == incoming_enrollment[:user_sourced_id]
+      end&.user
+      quill_classroom = existing_provider_classrooms.find do |ec|
+        ec.external_id == incoming_enrollment[:classroom_sourced_id]
+      end&.classroom
+      { quill_user:, quill_classroom: }
+    end
+
+    def run
+      user_and_classroom => { quill_user:, quill_classroom: }
```

Comment:
I think pattern matching here is kind of confusing.  

Instead, you could introduce two separate methods:

```
def quill_user
   @quill_user ||= 
      existing_provider_users
        .find {  |eu|. ec.external_id == incoming_enrollment[:user_sourced_id] }
        &.user
end
```

```
def quill_classroom
  @quill_classroom ||=
      existing_provider_classrooms
        .find { |ec| ec.external_id == incoming_enrollment[:classroom_sourced_id] }
        &.classroom
end
```

then you're the beginning of `run` becomes simplified:

```
def run
  return unless quill_classroom && quill_user
```